src/collision.cpp

```diff
@@ -455,95 +469,120 @@ collisionMoveResult collisionMoveSimple(Environment *env, IGameDef *gamedef,
+			float extra_dtime = nearest_dtime + 0.1f * fabsf(dtime - nearest_dtime);
+			stepbox.MinEdge.X += aspeed_f.X * extra_dtime;
+			stepbox.MinEdge.Z += aspeed_f.Z * extra_dtime;
+			stepbox.MaxEdge.X += aspeed_f.X * extra_dtime;
+			stepbox.MaxEdge.Z += aspeed_f.Z * extra_dtime;
 			// Check for stairs.
-			bool step_up = (nearest_collided != COLLISION_AXIS_Y) && // must not be Y direction
-					(movingbox.MinEdge.Y < cbox.MaxEdge.Y) &&
-					(movingbox.MinEdge.Y + stepheight > cbox.MaxEdge.Y) &&
-					(!wouldCollideWithCeiling(cinfo, stepbox,
-							cbox.MaxEdge.Y - movingbox.MinEdge.Y,
-							d));
-
-			// Get bounce multiplier
-			float bounce = -(float)nearest_info.bouncy / 100.0f;
-
-			// Move to the point of collision and reduce dtime by nearest_dtime
-			if (nearest_dtime < 0) {
-				// Handle negative nearest_dtime
-				if (!step_up) {
-					if (nearest_collided == COLLISION_AXIS_X)
-						pos_f->X += speed_f->X * nearest_dtime;
-					if (nearest_collided == COLLISION_AXIS_Y)
-						pos_f->Y += speed_f->Y * nearest_dtime;
-					if (nearest_collided == COLLISION_AXIS_Z)
-						pos_f->Z += speed_f->Z * nearest_dtime;
+			step_up = (movingbox.MinEdge.Y < cbox.MaxEdge.Y) &&
+				(movingbox.MinEdge.Y + stepheight > cbox.MaxEdge.Y) &&
+				(!wouldCollideWithCeiling(cinfo, stepbox,
+						cbox.MaxEdge.Y - movingbox.MinEdge.Y,
+						d));
+		}
+
+		// Get bounce multiplier
+		float bounce = -(float)nearest_info.bouncy / 100.0f;
+
+		// Move to the point of collision and reduce dtime by nearest_dtime
+		if (nearest_dtime < 0) {
+			// Handle negative nearest_dtime
+			// This largely means an "instant" collision, e.g., with the floor.
+			// We use aspeed and nearest_dtime to be consistent with above and resolve this collision
+			if (!step_up) {
+				if (nearest_collided == COLLISION_AXIS_X)
+					pos_f->X += aspeed_f.X * nearest_dtime;
+				if (nearest_collided == COLLISION_AXIS_Y)
+					pos_f->Y += aspeed_f.Y * nearest_dtime;
+				if (nearest_collided == COLLISION_AXIS_Z)
+					pos_f->Z += aspeed_f.Z * nearest_dtime;
+			}
+		} else if (nearest_dtime > 0) {
+			// updated average speed for the sub-interval up to nearest_dtime
+			aspeed_f = *speed_f + accel_f * 0.5f * nearest_dtime;
+			*pos_f += aspeed_f * nearest_dtime;
+			// Speed at (approximated) collision:
+			*speed_f += accel_f * nearest_dtime;
+			// Limit speed for avoiding hangs
+			*speed_f = truncate(rangelimv(*speed_f, -5000.0f, 5000.0f), 10000.0f);
+			dtime -= nearest_dtime;
+		}
+
+		v3f old_speed_f = *speed_f;
+
+		// Set the speed component that caused the collision to zero
+		if (step_up) {
+			// Special case: Handle stairs
+			nearest_info.is_step_up = true;
+		} else if (nearest_collided == COLLISION_AXIS_X) {
+			if (bounce < -1e-4 && fabsf(speed_f->X) > BS * 3) {
+				speed_f->X *= bounce;
+			} else {
+				speed_f->X = 0;
+				accel_f.X = 0; // avoid colliding in the next interations
+			}
+		} else if (nearest_collided == COLLISION_AXIS_Y) {
```

Comment:
Shouldn't this collision occur once in every function call when standing on ground?
Currently `touching_ground` is mostly set by the `std::fabs(cbox.MaxEdge.Y - box.MinEdge.Y) < 0.05f` condition further below. `if (speed_f->Y < 0.0f)` is only true when falling (can be tested by holding Jump while falling and touching ground).

Reply:
Well, if the computation is exact, then the collision happens at time 0 with speed 0; so a `<= 0.0f` may be enough.